libyul/optimiser/RedundantStoreEliminator.h

```diff
@@ -0,0 +1,113 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Optimiser component that removes stores to memory and storage slots that are not used
+ * or overwritten later on.
+ */
+
+#pragma once
+
+#include <libyul/ASTForward.h>
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/optimiser/RedundantStoreBase.h>
+
+#include <map>
+#include <vector>
+
+namespace solidity::yul
+{
+struct Dialect;
+struct AssignedValue;
+
+/**
+ * Optimizer component that removes mstore, sstore and similar statements if they
+ * are overwritten in all code paths or never read from.
+ *
+ * The m_store member of RedundantStoreBase is only used with the empty yul string
+ * as key in the first dimension.
+ */
+class RedundantStoreEliminator: public RedundantStoreBase
+{
+public:
+	static constexpr char const* name{"RedundantStoreEliminator"};
+	static void run(OptimiserStepContext& _context, Block& _ast);
+
+	explicit RedundantStoreEliminator(
+		Dialect const& _dialect,
+		std::map<YulString, SideEffects> const& _functionSideEffects,
+		std::map<YulString, AssignedValue> const& _ssaValues,
+		bool _ignoreMemory
+	):
+		RedundantStoreBase(_dialect),
+		m_ignoreMemory(_ignoreMemory),
+		m_functionSideEffects(_functionSideEffects),
+		m_ssaValues(_ssaValues)
+	{}
+
+	using RedundantStoreBase::operator();
+	void operator()(FunctionCall const& _functionCall) override;
+	void operator()(FunctionDefinition const&) override;
+	void operator()(Leave const&) override;
+
+	using RedundantStoreBase::visit;
+	void visit(Statement const& _statement) override;
+
+	enum class Location { Storage, Memory };
+	enum class Effect { Read, Write };
+	struct Operation
+	{
+		Location location;
+		Effect effect;
+		/// Start of affected area. Unknown if not provided.
+		std::optional<YulString> start;
+		/// Length of affected area, unknown if not provided.
+		/// Unused for storage.
+		std::optional<YulString> length;
+	};
+
+private:
+	void shortcutNestedLoop(TrackedStores const&) override
+	{
+		// We might only need to do this for newly introduced stores in the loop.
+		changeUndecidedTo(State::Used);
+	}
+	void finalizeFunctionDefinition(FunctionDefinition const&) override { scheduleUnusedForDeletion(); }
+
+
+	std::vector<Operation> operationsFromFunctionCall(FunctionCall const& _functionCall) const;
+	void applyOperation(Operation const& _operation);
+	bool knownUnrelated(Operation const& _op1, Operation const& _op2) const;
+	bool knownCovered(Operation const& _covered, Operation const& _covering) const;
+
+	using TrackedStores = std::map<ExpressionStatement const*, State>;
+
+	void changeUndecidedTo(State _newState, std::optional<Location> _onlyLocation = std::nullopt);
+	void scheduleUnusedForDeletion();
+
+	std::optional<YulString> identifierIfSSA(Expression const& _expression) const;
```

Comment:
```suggestion
	std::optional<YulString> identifierNameIfSSA(Expression const& _expression) const;
```
Hopefully not too verbose?